[JSC] add support for `Uint8Array.fromHex` and `Uint8Array.prototype.setFromHex`

dcrousso opened the pull request:
#### eeedf04c47ffec014748b4d743f31a68f2a3f292
<pre>
[JSC] add support for `Uint8Array.fromHex` and `Uint8Array.prototype.setFromHex`
<a href="https://bugs.webkit.org/show_bug.cgi?id=275593">https://bugs.webkit.org/show_bug.cgi?id=275593</a>

Reviewed by Yusuke Suzuki.

These methods will make it easier for developers to decode typed character/byte data.

Note that there are other methods in the related proposal, but for ease of reviewing they will be implemented separately.

Spec: &lt;<a href="https://tc39.es/proposal-arraybuffer-base64/spec/">https://tc39.es/proposal-arraybuffer-base64/spec/</a>&gt;
Proposal: &lt;<a href="https://github.com/tc39/proposal-arraybuffer-base64">https://github.com/tc39/proposal-arraybuffer-base64</a>&gt;

* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewConstructor.h:
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewConstructorInlines.h:
(JSC::JSGenericTypedArrayViewConstructor&lt;ViewClass&gt;::finishCreation):
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewConstructor.cpp: Added.
(JSC::uint8ArrayConstructorFromHex):

* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototype.h:
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeInlines.h:
(JSC::JSGenericTypedArrayViewPrototype&lt;ViewClass&gt;::finishCreation):
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototype.cpp:
(JSC::uint8ArrayPrototypeSetFromHex):

* Source/JavaScriptCore/runtime/CommonIdentifiers.h:
Create builtin identifiers for `read` and `written`.

* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:

* JSTests/stress/uint8array-fromHex.js: Added.
* JSTests/stress/uint8array-setFromHex.js: Added.

Canonical link: <a href="https://commits.webkit.org/280967@main">https://commits.webkit.org/280967@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/47657272cc06613b6e6367cba2afb7bc6edb684a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57824 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37152 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10300 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61446 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8269 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44788 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8457 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46853 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5870 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59854 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34845 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49972 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27681 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31613 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7265 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7273 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/50916 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/53561 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7535 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63129 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/57066 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1738 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7606 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54073 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1744 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49983 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54194 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12867 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1478 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/78827 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32981 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/13078 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34067 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35151 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33812 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/JSGenericTypedArrayViewConstructor.cpp:
```diff
@@ -0,0 +1,65 @@
+    Structure* structure = globalObject->typedArrayStructure(TypeUint8, false);
+    JSUint8Array* uint8Array = JSUint8Array::create(vm, structure, Uint8Array::create(bytes.span()));
+    RETURN_IF_EXCEPTION(scope, { });
+    RELEASE_AND_RETURN(scope, JSValue::encode(uint8Array));
```
Just do,
```
return JSUint8Array::create(vm, globalObject->typedArrayStructure(TypeUint8, false), Uint8Array::create(bytes.span()));
```

Constellation replied:
Instead of allocating Vector and copying them to the typed array, we should allocate uninitialized typed array and fill it with values.

Constellation commented on Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototype.cpp:
```diff
@@ -31,6 +31,45 @@
+    resultObject->putDirect(vm, vm.propertyNames->read, jsNumber(read));
+    resultObject->putDirect(vm, vm.propertyNames->written, jsNumber(written));
+
+    RELEASE_AND_RETURN(scope, JSValue::encode(resultObject));
```
We do not need this.

dcrousso replied:
do you mind explaining when one would want to use `RELEASE_AND_RETURN`?

Constellation commented on Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototype.cpp:
```diff
@@ -31,6 +31,45 @@
+    }
+
+    JSObject* resultObject = constructEmptyObject(globalObject);
+    RETURN_IF_EXCEPTION(scope, { });
```
Ditto.

dcrousso replied:
do you mind explaining when it's necessary (or not) to use `RETURN_IF_EXCEPTION`?  is there a general rule (i.e. only certain functions will throw an exception)?

Constellation commented on Source/JavaScriptCore/runtime/JSGenericTypedArrayViewConstructor.cpp:
```diff
@@ -0,0 +1,62 @@
+        int ones = parseDigit(view[read++], 16);
+        if (UNLIKELY(tens < 0 || tens > 255 || ones < 0 || ones > 255))
+            return JSValue::encode(throwSyntaxError(globalObject, scope, "Uint8Array.fromHex requires a string containing only \"0123456789abcdefABCDEF\""_s));
+        uint8array->typedVector()[written++] = (tens * 16) + ones;
```
Let's just call `typedVector` outside of this loop and use the data pointer.

Constellation commented on Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototype.cpp:
```diff
@@ -31,6 +31,43 @@
+
+    size_t read = 0;
+    size_t written = 0;
+    while (read < view.length() && written < uint8Array->length()) {
```
We can compute how many iterations we need to do based on uint8Array size limit and view's length. Let's compute this and do iteration with that count.

Constellation reviewed (approved):
r=me

dcrousso committed:
[JSC] add support for `Uint8Array.fromHex` and `Uint8Array.prototype.setFromHex`
https://bugs.webkit.org/show_bug.cgi?id=275593

Reviewed by Yusuke Suzuki.

These methods will make it easier for developers to decode typed character/byte data.

Note that there are other methods in the related proposal, but for ease of reviewing they will be implemented separately.

Spec: <https://tc39.es/proposal-arraybuffer-base64/spec/>
Proposal: <https://github.com/tc39/proposal-arraybuffer-base64>

* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewConstructor.h:
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewConstructorInlines.h:
(JSC::JSGenericTypedArrayViewConstructor<ViewClass>::finishCreation):
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewConstructor.cpp: Added.
(JSC::uint8ArrayConstructorFromHex):

* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototype.h:
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeInlines.h:
(JSC::JSGenericTypedArrayViewPrototype<ViewClass>::finishCreation):
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototype.cpp:
(JSC::uint8ArrayPrototypeSetFromHex):

* Source/JavaScriptCore/runtime/CommonIdentifiers.h:
Create builtin identifiers for `read` and `written`.

* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:

* JSTests/stress/uint8array-fromHex.js: Added.
* JSTests/stress/uint8array-setFromHex.js: Added.

Canonical link: https://commits.webkit.org/280967@main
